src/github.com/matrix-org/dendrite/roomserver/api/output.go

```diff
@@ -41,6 +45,45 @@ type OutputEvent struct {
 	NewInviteEvent *OutputNewInviteEvent `json:"new_invite_event,omitempty"`
 	// The content of event with type OutputTypeRetireInviteEvent
 	RetireInviteEvent *OutputRetireInviteEvent `json:"retire_invite_event,omitempty"`
+	// Serialized span context
+	OpentracingCarrier opentracing.TextMapCarrier `json:"opentracing_carrier"`
+}
+
+// AddSpanFromContext fills out the OpentracingCarrier field from the given context
+func (o *OutputEvent) AddSpanFromContext(ctx context.Context) error {
+	span := opentracing.SpanFromContext(ctx)
+	ext.SpanKindProducer.Set(span)
+	carrier := make(opentracing.TextMapCarrier)
+	tracer := opentracing.GlobalTracer()
+
+	err := tracer.Inject(span.Context(), opentracing.TextMap, carrier)
+	if err != nil {
+		return err
+	}
+
+	o.OpentracingCarrier = carrier
+
+	return nil
+}
+
+// StartSpanAndReplaceContext produces a context and opentracing span from the
+// info embedded in OutputEvent
+func (o *OutputEvent) StartSpanAndReplaceContext(
```

Comment:
I'm not sure it's fair to expect people to realise that "Span" refers to opentracing; ideally this name would make it clear, but I appreciate `StartOpentracingSpanAndReplaceContext` is not a catchy name. I'd like us to think about better names, but perhaps I need to wait for some clarity on how this fits together